mixer/pkg/protobuf/yaml/encoder_test.go

```diff
@@ -0,0 +1,287 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// nolint
+//go:generate protoc testdata/all/types.proto -otestdata/all/types.descriptor -I$GOPATH/src/istio.io/istio/vendor/istio.io/api -I.
+//go:generate $GOPATH/src/istio.io/istio/bin/mixer_codegen.sh -f mixer/pkg/protobuf/yaml/testdata/all/types.proto
+
+package yaml
+
+import (
+	"bytes"
+	"io/ioutil"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/ghodss/yaml"
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/protoc-gen-gogo/descriptor"
+	yaml2 "gopkg.in/yaml.v2"
+
+	"istio.io/istio/mixer/pkg/protobuf/yaml/testdata/all"
+)
+
+var (
+	empty = ``
+
+	unknownField = `
+fxxx: 123
+`
+	simple = `
+str: "mystring"
+i64: 56789
+dbl: 123.456
+b: true
+enm: TWO
+oth:
+  str: "mystring2"
+  i64: 33333
+  dbl: 333.333
+  b: true
+  inenum: INNERTHREE
+  inmsg:
+    str: "myinnerstring"
+    i64: 99
+    dbl: 99.99
+`
+	inner = `
+str: "mystring"
+i64: 56789
+dbl: 123.456
+b: true
+`
+
+	badStr = `
+str: 12345
+`
+	badBool = `
+b: ""
+`
+	badInt64 = `
+i64: ""
+`
+
+	badInt64SubMsg = `
+oth:
+  i64: ""
+`
+
+	badDouble = `
+dbl: ""
+`
+	badEnumType = `
+enm: 12345
+`
+	badEnumValue = `
+enm: BADNAME
+`
+	badMessage = `
+oth: "string instead of message"
+`
+	// Not supported yet; will be added soon.
+	unsupportedType = `
+byts: "abcd"
+`
+)
+
+type testdata struct {
+	n           string
+	input       string
+	msg         string
+	skipUnknown bool
+	err         string
+	expected    proto.Message
+}
+
+func TestEmptyMessage(t *testing.T) {
+	fds, err := getFileDescSet("testdata/all/types.descriptor")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, td := range []testdata{
+		{
```

Comment:
We should add a testcase that tests `Simple.map_str_str`.
`StringMaps` is our mainstream use case.